8345259: Disallow ALL-MODULE-PATH without explicit --module-path

jerboaa opened the pull request:
Please review this extension to #22609 which now disallows `ALL-MODULE-PATH` without explicit `--module-path` option or a non-existent module path. In addition, this fixes a bug mentioned in #22609 when `ALL-MODULE-PATH` and `--limit-modules` are used in combination. It failed earlier and passes now due to alignment of `ModuleFinder`s. With this patch JEP 493 enabled builds and regular JDK builds behave the same in terms of `ALL-MODULE-PATH`.

When an explicit module path is being added, there is no difference. All modules on that path will be added as roots. Tests have been added for the various cases and existing tests updated to allow for them to run on JEP 493 enabled builds. Thoughts?

Testing:
- [x] GHA, `test/jdk/tools/jlink` (all pass)
- [x] Added jlink test.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8345894](https://bugs.openjdk.org/browse/JDK-8345894) to be approved

### Issues
 * [JDK-8345259](https://bugs.openjdk.org/browse/JDK-8345259): Disallow ALL-MODULE-PATH without explicit --module-path (**Bug** - P3)
 * [JDK-8345894](https://bugs.openjdk.org/browse/JDK-8345894): Disallow ALL-MODULE-PATH without explicit --module-path (**CSR**)


### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22494/head:pull/22494` \
`$ git checkout pull/22494`

Update a local copy of the PR: \
`$ git checkout pull/22494` \
`$ git pull https://git.openjdk.org/jdk.git pull/22494/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22494`

View PR using the GUI difftool: \
`$ git pr show -t 22494`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22494.diff">https://git.openjdk.org/jdk/pull/22494.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22494#issuecomment-2512379345)
</details>

jerboaa commented on test/jdk/tools/jlink/runtimeImage/BasicJlinkTest.java:
```diff
@@ -39,7 +39,7 @@
  * @build tests.* jdk.test.lib.process.OutputAnalyzer
  *        jdk.test.lib.process.ProcessTools
- * @run main/othervm -Xmx1400m BasicJlinkTest false
+ * @run main/othervm -Xmx1400m BasicJlinkTest
```
Note to reviewers: This is just a clean-up of a parameter that isn't being used, so I've removed it here.

AlanBateman commented:
I'm in two minds as to whether we should do anything here.

ALL-MODULE-PATH means all observable modules on the module path but here, it's all observable modules minus jdk.jlink or any module that directly requires jdk.jlink. This is hard to reason about, and not exactly right either because any filtering should be any module that transitively requires jdk.jlink.

As background, the original motivation for ALL-MODULE-PATH was Maven build. It has always been supported by jlink too but I can't recall, or find, why we did that. Maybe Mandy can remember. In JEP 261, it is listed as being valid for both compile-time and run-time, but also mentions link-time.

mlchung commented:
> ALL-MODULE-PATH means all observable modules on the module path but here, it's all observable modules minus jdk.jlink or any module that directly requires jdk.jlink. This is hard to reason about, and not exactly right either because any filtering should be any module that transitively requires jdk.jlink.

I have the same comment.  ALL-MODULE-PATH means all observable modules on the module path that includes `jdk.jlink`.  It would be confusing if the resulting image linked from the run-time image is different than the image linked from packaged modules.   I think it may just be another restriction that `--add-modules ALL-MODULE-PATH` can't be used when linking from the run-time image for now.

> As background, the original motivation for ALL-MODULE-PATH was Maven build. It has always been supported by jlink too but I can't recall, or find, why we did that. Maybe Mandy can remember. In JEP 261, it is listed as being valid for both compile-time and run-time, but also mentions link-time.

I don't recall neither.   It might just be implemented as consistent with java and javac.

jerboaa commented:
> > ALL-MODULE-PATH means all observable modules on the module path but here, it's all observable modules minus jdk.jlink or any module that directly requires jdk.jlink. This is hard to reason about, and not exactly right either because any filtering should be any module that transitively requires jdk.jlink.
> 
> I have the same comment. ALL-MODULE-PATH means all observable modules on the module path that includes `jdk.jlink`. It would be confusing if the resulting image linked from the run-time image is different than the image linked from packaged modules. I think it may just be another restriction that `--add-modules ALL-MODULE-PATH` can't be used when linking from the run-time image for now and removed together when the other restriction is removed.

That makes sense. As long as we handle it, I'm fine. So my thinking then is:

- Abort the link with an appropriate error when `ALL-MODULE-PATH` is being used an link from the run-time image is used but no module path given (`--module-path` arg missing).
- Continue as is now when linking from the run-time image and `--module-path` arg is given (there is no difference to a run with packaged modules).

Does that sound OK to you?

jerboaa commented:
> > Abort the link with an appropriate error when ALL-MODULE-PATH is being used and a link from the run-time image is used but no module path given (--module-path arg missing).
> 
> Just to double check, with the default build, then `jlink --add-modules ALL-MODULE-PATH --output myimage` would fail because the module path hasn't been specified. No difference between default and --generate-linkable-runtime build. It would be a change in behaviour but I think the right thing to do.

Originally I hadn't intended to do that, but I can add this. Should I do it in this bug or a separate one? I would think that this change of behaviour would need a CSR?

AlanBateman commented:
> > Just to double check, with the default build, then `jlink --add-modules ALL-MODULE-PATH --output myimage` would fail because the module path hasn't been specified. No difference between default and --generate-linkable-runtime build. It would be a change in behaviour but I think the right thing to do.
> 
> Originally I hadn't intended to do that, but I can add this. Should I do it in this bug or a separate one? I would think that this change of behaviour would need a CSR?

I think this case was missed when jlink was modified to default the module path to $JDK_HOME/jmods. I think my preference would be for this case to fail when --module-path is not specified. Let's see if Mandy agrees. If we do this change then it would be a behavior change so would need a CSR.

mlchung commented:
Such behavioral change is a good change as jlink from the default and --generate-linkable-runtime build would have the consistent behavior.   If a module path is given with no root module (empty path), it throws the following error.  I think it can throw a similar message as if `--add-modules ALL-MODULE-PATH` is given but no `--module-path`.

```
$ jlink --add-modules ALL-MODULE-PATH --output myimage --module-path emptyPath
Error: Cannot invoke "java.nio.file.Path.getFileName()" because "javaBasePath" is null
java.lang.NullPointerException: Cannot invoke "java.nio.file.Path.getFileName()" because "javaBasePath" is null
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.isJavaBaseFromDefaultModulePath(JlinkTask.java:660)
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.targetPlatform(JlinkTask.java:632)
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.createImageProvider(JlinkTask.java:569)
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.createImage(JlinkTask.java:410)
	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.run(JlinkTask.java:285)
	at jdk.jlink/jdk.tools.jlink.internal.Main.run(Main.java:56)
	at jdk.jlink/jdk.tools.jlink.internal.Main.main(Main.java:34)
```

jerboaa commented:
OK. I'll update the PR accordingly.

jerboaa commented:
/csr needed

jerboaa commented:
> Such behavioral change is a good change as jlink from the default and --generate-linkable-runtime build would have the consistent behavior. If a module path is given with no root module (empty path), it throws the following error. I think it can throw a similar message as if `--add-modules ALL-MODULE-PATH` is given but no `--module-path`.
> 
> ```
> $ jlink --add-modules ALL-MODULE-PATH --output myimage --module-path emptyPath
> Error: Cannot invoke "java.nio.file.Path.getFileName()" because "javaBasePath" is null
> java.lang.NullPointerException: Cannot invoke "java.nio.file.Path.getFileName()" because "javaBasePath" is null
> 	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.isJavaBaseFromDefaultModulePath(JlinkTask.java:660)
> 	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.targetPlatform(JlinkTask.java:632)
> 	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.createImageProvider(JlinkTask.java:569)
> 	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.createImage(JlinkTask.java:410)
> 	at jdk.jlink/jdk.tools.jlink.internal.JlinkTask.run(JlinkTask.java:285)
> 	at jdk.jlink/jdk.tools.jlink.internal.Main.run(Main.java:56)
> 	at jdk.jlink/jdk.tools.jlink.internal.Main.main(Main.java:34)
> ```

Thanks, yes. Added.

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -397,7 +399,22 @@ private JlinkConfiguration initJlinkConfig() throws BadArgs {
         for (String mod : options.addMods) {
             if (mod.equals(ALL_MODULE_PATH) && options.modulePath.size() > 0) {
-                ModuleFinder mf = newModuleFinder(rootsFinder, options.limitMods, Set.of(), isLinkFromRuntime);
+                // Apply a module limit for the roots finder if it was otherwise empty.
+                // Since we are using the same finder for determining the roots set as
+                // for the actual link, we need to apply this trick so as to not
+                // include all modules of the JDK plus the extra module path in
+                // the roots set.
+                Set<String> allModsLimits = options.limitMods;
+                if (options.limitMods.isEmpty()) {
+                    Set<String> modsLimits = new HashSet<>();
+                    initialFinder.findAll()
+                                 .stream()
+                                 .map(ModuleReference::descriptor)
+                                 .map(ModuleDescriptor::name)
+                                 .forEach(mn -> modsLimits.add(mn));
+                    allModsLimits = modsLimits;
+                }
+                ModuleFinder mf = newModuleFinder(rootsFinder, allModsLimits, Set.of(), isLinkFromRuntime);
```
It took me some time to understand this change as it depends on https://github.com/openjdk/jdk/pull/22609.   I adjusted the code trying to help the readers easier to understand.   

See what you think.   Instead of the `determineLinkFromRuntime` method, it does the check in this method that makes it explicit to the readers.

```
   private JlinkConfiguration initJlinkConfig() throws BadArgs {
        // Empty module path not allowed with ALL-MODULE-PATH in --add-modules
        if (options.addMods.contains(ALL_MODULE_PATH) && options.modulePath.isEmpty()) {
            throw taskHelper.newBadArgs("err.all.module.path.empty.mod.path");
        }

        ModuleFinder appModulePathFinder = createFinderFromPath(options.modulePath);
        ModuleFinder finder = appModulePathFinder;
        boolean isLinkFromRuntime = false;
        // if packaged module for java.base is not found, either from
        // the default module path or the run-time image
        if (!appModulePathFinder.find("java.base").isPresent()) {
            // Add the default module path if that exists
            Path defModPath = getDefaultModulePath();
            if (defModPath != null) {
                options.modulePath.add(defModPath);
                finder = createFinderFromPath(options.modulePath);
            }
            if (!finder.find("java.base").isPresent()) {
                isLinkFromRuntime = true;
                // java.base is from the system module path
                finder = ModuleFinder.compose(ModuleFinder.ofSystem(), appModulePathFinder);
            }
        }

        // Determine the roots set
        Set<String> roots = new HashSet<>();
        for (String mod : options.addMods) {
            if (mod.equals(ALL_MODULE_PATH)) {
                // all observable modules are roots
                Set<String> initialRoots = appModulePathFinder.findAll()
                        .stream()
                        .map(ModuleReference::descriptor)
                        .map(ModuleDescriptor::name)
                        .collect(Collectors.toSet());

                // Error if no module is found on module path
                if (initialRoots.isEmpty()) {
                    throw taskHelper.newBadArgs("err.all.module.path.empty.mod.path");
                }

                // Use a module finder with limited observability to find the observable modules
                // as not all JDK modules on the default module path or the run-time image
                ModuleFinder mf = newModuleFinder(finder, options.limitMods.isEmpty() ? initialRoots : options.limitMods,
                                                  Set.of(), isLinkFromRuntime);
                mf.findAll()
                  .stream()
                  .map(ModuleReference::descriptor)
                  .map(ModuleDescriptor::name)
                  .forEach(mn -> roots.add(mn));
            } else {
                roots.add(mod);
            }
        }
        finder = newModuleFinder(finder, options.limitMods, roots, isLinkFromRuntime);

        // --keep-packaged-modules doesn't make sense as we are not linking
        // from packaged modules to begin with.
        if (isLinkFromRuntime && options.packagedModulesPath != null) {
            throw taskHelper.newBadArgs("err.runtime.link.packaged.mods");
        }

        return new JlinkConfiguration(options.output,
                                      roots,
                                      finder,
                                      isLinkFromRuntime,
                                      options.ignoreModifiedRuntime,
                                      options.generateLinkableRuntime);
    }
```

mlchung replied:
Is the `isLinkFromRuntime` parameter needed in the `newModuleFinder` method?   The java.base version check does not have any issue for linking from run-time image, right?

jerboaa replied:
> Is the `isLinkFromRuntime` parameter needed in the `newModuleFinder` method? The java.base version check does not have any issue for linking from run-time image, right?

It's not needed other than for skipping the check (as it'll trivially pass). I can remove it if that's preferred.

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/resources/jlink.properties:
```diff
@@ -127,6 +127,7 @@ err.runtime.link.packaged.mods=This JDK has no packaged modules.\
 err.runtime.link.modified.file={0} has been modified
 err.runtime.link.patched.module=jlink does not support linking from the run-time image\
 \ when running on a patched runtime with --patch-module
+err.all.module.path.empty.mod.path=ALL-MODULE-PATH requires --module-path option (or --module-path does not exist)
 err.empty.module.path=empty module path
```
`err.empty.module.path` is unused.

```suggestion
err.no.module.path=--module-path option must be specified with --add-modules ALL-MODULE-PATH 
err.empty.module.path=No module found from the module path {0}
```
```

jerboaa replied:
I've updated it and mentioned `ALL-MODULE-PATH` in the error of a non-existing module path.

mlchung commented on test/jdk/tools/jlink/basic/AllModulePathTest.java:
```diff
@@ -0,0 +1,200 @@
+ *          jdk.jlink/jdk.tools.jimage
+ * @run main/othervm -Xmx1g AllModulePathTest
+ */
+public class AllModulePathTest {
```
Any reason why these test cases can't be added to `AllModulePath`?  `AllModulePathTest` and `AllModulePath` test names are very alike.

jerboaa replied:
I'll merge the tests into `AllModulePath`.

mlchung reviewed (commented):
This comment is also part of https://github.com/openjdk/jdk/pull/22609.

mlchung commented on test/jdk/tools/jlink/basic/AllModulePath.java:
```diff
@@ -153,6 +290,23 @@ private void checkModules(Path image, Set<String> modules) throws Throwable {
+     */
+    private void verifyListModules(Path targetPath, List<String> expected) throws Exception {
+        Path java = findTool(targetPath, "java");
+        List<String> listMods = List.of(java.toString(), "--list-modules");
+        OutputAnalyzer out = ProcessTools.executeCommand(listMods.toArray(new String[] {}));
+        if (out.getExitValue() != 0) {
+            throw new AssertionError("java --list-modules failed");
+        }
+        List<String> actual = Stream.of(out.getStdout().split(Pattern.quote(System.lineSeparator())))
+                                    .map(s -> { return s.split("@")[0]; })
+                                    .sorted()
+                                    .toList();
```
```suggestion
        OutputAnalyzer out = ProcessTools.executeCommand(java.toString(), "--list-modules");
                                         .shouldHaveExitValue(0);
        List<String> actual = out.asLines().stream()
                                 .map(s -> { return s.split("@")[0]; })
                                 .sorted()
                                 .toList();
```

mlchung commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -419,8 +417,30 @@ private JlinkConfiguration initJlinkConfig() throws BadArgs {
+                // not include JDK modules from the default module path or the
+                // run-time image. Only do this if no --limit-modules has been
+                // specified to begin with.
+                ModuleFinder mf = newLimitedFinder(finder,
```
FWIW, rereading this, it's still a bit awk to call a module finder a limited finder.

jerboaa replied:
How about `newScopedFinder()`?

mlchung replied:
Best to leave out renaming when in doubt - either `newModuleFinder` or `limitFinder`.

jerboaa replied:
OK. I'll go with `limitFinder`.

mlchung reviewed (commented):
Need to bump the copyright end year in the tests.

mlchung commented on test/jdk/tools/jlink/basic/AllModulePath.java:
```diff
@@ -137,7 +164,120 @@ public void testAddModules() throws Throwable {
+     * limits applied.
+     */
+    @Test
+    public void modulePathWithLimitMods() throws Exception {
```
Is this test case different than `testLimitModules` which does
```
jlink --output image1 --add-modules ALL-MODULE-PATH --module-path mods --limit-modules m1
```

`testAddModules` also tests:

```
jlink --output image2 --add-modules m1,test --add-modules ALL-MODULE-PATH --module-path mods --limit-modules java.base
```

jerboaa replied:
We have:
- `testLimitModules()` limits on a module from the module path, `m1`. It expects `m1` and it's dependencies in the image.
- `testAddModules()` limits on a module dependency, but has modules from the module path added explicitly: `m1` and `test`. Therefore, expects `m1` and `test` as well as there deps in the image.
- `modulePathWithLimitMods()` limits on a module dependency, but doesn't include any other module with the `--add-modules` option. Therefore, only expects `jdk.jfr` and its dependency, `java.base` to be included in the image.

They are all slightly different and it's useful to test all 3 combinations.

jerboaa replied:
I've tried to clarify the cases with comments.

mlchung replied:
Indeed useful tests.   This uncovers an issue about `--limit-modules`.  Thanks.

mlchung commented on test/jdk/tools/jlink/basic/AllModulePath.java:
```diff
@@ -137,7 +164,120 @@ public void testAddModules() throws Throwable {
+     * --add-modules ALL-MODULE-PATH with an existing module-path.
+     */
+    @Test
+    public void modulePath() throws Exception {
```
Any difference to `testAllModulePath` which does:

```
jlink --output image --add-modules ALL-MODULE-PATH --module-path mods
```

jerboaa replied:
No difference. I've dropped it.

jerboaa commented on test/jdk/tools/jlink/basic/AllModulePath.java:
```diff
@@ -44,35 +37,64 @@
+ * @build jdk.test.lib.process.ProcessTools
+ *        jdk.test.lib.process.OutputAnalyzer
+ *        jdk.test.lib.compiler.CompilerUtils
+ * @run testng/othervm -Duser.language=en -Duser.country=US AllModulePath
```
> Is `-Duser.language=en -Duser.country=US` needed?

I think so, since the test asserts the error message which is potentially localized.

> Please add `@bug` to this test.

Sure.

jerboaa commented on src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:
```diff
@@ -377,8 +369,14 @@ public static void createImage(JlinkConfiguration config,
+        if (options.addMods.contains(ALL_MODULE_PATH) && options.modulePath.isEmpty()) {
+            throw taskHelper.newBadArgs("err.no.module.path");
+        }
+        List<Path> originalModulePath = new ArrayList<>(options.modulePath);
```
Correct. Nice idea. Will fix.

jerboaa commented on test/jdk/tools/jlink/basic/AllModulePath.java:
```diff
@@ -137,7 +164,117 @@ public void testAddModules() throws Throwable {
+        if (isExplodedJDKImage()) {
+            return;
+        }
+        Path targetPath = HELPER.createNewImageDir("all-mod-path-no-mod-path");
+        List<String> allArgs = List.of("--add-modules", "ALL-MODULE-PATH",
+                                       "--output", targetPath.toString());
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        PrintWriter out = new PrintWriter(baos);
+        ByteArrayOutputStream berrOs = new ByteArrayOutputStream();
+        PrintWriter err = new PrintWriter(berrOs);
+        int rc = JLINK_TOOL.run(out, err, allArgs.toArray(new String[] {}));
+        assertTrue(rc != 0);
+        String actual = new String(baos.toByteArray()).trim();
+        assertEquals(actual, "Error: --module-path option must be specified with --add-modules ALL-MODULE-PATH");
```
Yes, agreed. It should be much nicer now.

mlchung commented on test/jdk/tools/jlink/basic/AllModulePath.java:
```diff
@@ -84,60 +102,141 @@ public void setup() throws Throwable {
+        JlinkOutput allOut = createImage(targetPath, allArgs, true /* success */);
+        assertTrue(allOut.stdout.isEmpty());
+        assertTrue(allOut.stderr.isEmpty());
+        List<String> expected = List.of("java.base", "jdk.jfr");
```
```
jlink --add-modules ALL-MODULE-PATH --limit-modules jdk.jfr --module-path jmods --output all-mods-limit-mods.image 
```

This PR changes to find the root modules from the paths specified in `--module-path` only.   So I would expect that `jdk.jfr` should not be included in the resulting image because it's not on `--module-path`.    This seems a bug as it finds all modules using the module finder that includes `jdk.jfr`.   The module finder should be:

```
                ModuleFinder mf = newLimitedFinder(finder,
                                                   options.limitMods.isEmpty() ? initialRoots : Set.of(),
                                                   Set.of());
```

jerboaa replied:
> This PR changes to find the root modules from the paths specified in --module-path only.

Yes, but in this case modules on the module path contain modules that require `jdk.jfr`:

```
$ jmod describe ./build/linux-x86_64-server-release/test-support/jtreg_test_jdk_tools_jlink_basic_AllModulePath_java/scratch/0/jmods/com.baz.runtime.jmod 
com.baz.runtime
exports com.baz.runtime
exports com.baz.runtime.com.foo.bar
requires java.base mandated
requires jdk.jfr
main-class com.baz.runtime.Main
```

So I think it's OK to limit to a smaller set of modules even though it seems a strange use-case.

A different question would be what should happen if the module specified in `--limit-modules` in **not** in the dependency graph of ALL-MODULE-PATH modules.

jerboaa replied:
> A different question would be what should happen if the module specified in `--limit-modules` is **not** in the dependency graph of ALL-MODULE-PATH modules.

This seems a good candidate for an error as it's unexpected, IMO. I'll add a separate test case which exercises this and add code to prevent this:

```
$ rm -rf ./build/limit-mods-test && ./build/linux-x86_64-server-release/images/jdk/bin/jlink --add-modules ALL-MODULE-PATH --module-path ./build/jmods --limit-modules jdk.net --verbose --output ./build/limit-mods-test
Error: jdk.net not in module dependency graph, but specified with --limit-modules
```

The same without limit modules:
```
$ rm -rf ./build/limit-mods-test && ./build/linux-x86_64-server-release/images/jdk/bin/jlink --add-modules ALL-MODULE-PATH --module-path ./build/jmods --verbose --output ./build/limit-mods-test
Linking based on the current run-time image
com.bar.testmod file:///disk/openjdk/upstream-sources/git/jdk-jdk/./build/jmods/com.bar.testmod.jmod
java.base jrt:/java.base (run-time image)
java.compiler jrt:/java.compiler (run-time image)
jdk.compiler jrt:/jdk.compiler (run-time image)
jdk.internal.opt jrt:/jdk.internal.opt (run-time image)
jdk.jdeps jrt:/jdk.jdeps (run-time image)
jdk.zipfs jrt:/jdk.zipfs (run-time image)

Providers:
  jdk.compiler provides com.sun.tools.javac.platform.PlatformProvider used by jdk.compiler,jdk.jdeps
  java.base provides java.nio.file.spi.FileSystemProvider used by java.base
  jdk.zipfs provides java.nio.file.spi.FileSystemProvider used by java.base
  jdk.compiler provides java.util.spi.ToolProvider used by java.base
  jdk.jdeps provides java.util.spi.ToolProvider used by java.base
  jdk.compiler provides javax.tools.JavaCompiler used by java.compiler
  jdk.compiler provides javax.tools.Tool not used by any observable module
```

A similar - pre-existing - problem exists where `--limit-modules` would be outside the dependency graph of modules specified with `--add-modules`. But this would be for a different bug so as to prevent scope-creep.

jerboaa replied:
> But this would be for a different bug so as to prevent scope-creep.

Filed https://bugs.openjdk.org/browse/JDK-8346299 for this.

mlchung replied:
```
jlink --add-modules ALL-MODULE-PATH --limit-modules jdk.jfr --module-path jmods --output all-mods-limit-mods.image
```

The set of observable modules only has `java.base` and `jdk.jfr` because of `--limit-modules`.   Hence `com.baz.runtime` from `jmods` is not observable.    `ALL-MODULE-PATH` adds all observable modules found on the relevant module paths are added to the root set.   This matches the run-time behavior.

```
$ java --list-modules -p com.baz.runtime.jar --add-modules ALL-MODULE-PATH --limit-modules jdk.jfr
java.base@25-internal
jdk.jfr@25-internal
```

mlchung replied:
Limiting the observable modules is sometimes useful for testing/debugging and to reduce the number of modules resolved when the main module is the unnamed module.

@AlanBateman  Is `--limit-modules` needed at link time?

jerboaa replied:
IMO it's also useful to limit included modules when `--bind-services` is being used.

jerboaa replied:
@mlchung It's not clear if you think there is an action expected from myself. If there is, please clarify what. Thanks!

mlchung commented on test/jdk/tools/jlink/basic/AllModulePath.java:
```diff
@@ -153,6 +252,21 @@ private void checkModules(Path image, Set<String> modules) throws Throwable {
+    /*
+     * Verify linked modules using java --list-modules
+     */
+    private void verifyListModules(Path targetPath, List<String> expected) throws Exception {
```
```suggestion
    private void verifyListModules(Path targetPath, Set<String> expected) throws Exception {
```

Can simply use a Set to compare the result and  no need to sort `actual` from the output.

jerboaa replied:
OK.

AlanBateman commented:
Just on --add-modules ALL-MODULE-PATH with --limit-modules. This combination isn't implemented at run-time, I think we just missed it at the time. In practical terms it's not interesting of course but just mentioning it here because I don't think jlink with --add-modules ALL-MODULE-PATH and --limit-modules is interesting either.

The main thing to deal with here is --add-modules ALL-MODULE-PATH without --module-path so I'll focus on that part.

mlchung commented:
> @mlchung It's not clear if you think there is an action expected from myself. If there is, please clarify what. Thanks!

Sorry for not being clear. 

As I look into the behavior of `--add-modules ALL-MODULE-PATH` with `--limit-modules` more link time and run-time, I find some uncertainty and question if `--add-modules ALL-MODULE-PATH` with `--limit-modules` is needed/interesting.

Thanks @AlanBateman for clarifying that `--add-modules ALL-MODULE-PATH` with `--limit-modules` combination isn't implemented at run-time.   While I agree that this is a separate issue, this PR depends on how this should be implemented because the test cases (testLimitedModules and testAddModules in AllModulePath.java) do not disambiguate the expected behavior.    Unsupporting this combination could be one option but different from run-time which I think it's not ideal.   I propose to implement as described in https://openjdk.org/jeps/261#Limiting-the-observable-modules -- to limit the observable modules to those in the transitive closure of the named modules plus the modules specified via the `--add-modules` option.   In other words, `--add-modules ALL-MODULE-PATH` adds all modules on module path to the observable set.   It's easy to explain and understand and no difference when `--limit-modules` is combined with `--add-modules m1,m2,...` with named modules.

jerboaa commented:
> I propose to implement as described in https://openjdk.org/jeps/261#Limiting-the-observable-modules -- to limit the observable modules to those in the transitive closure of the named modules plus the modules specified via the `--add-modules` option. In other words, `--add-modules ALL-MODULE-PATH` adds all modules on module path to the observable set. It's easy to explain and understand and no difference when `--limit-modules` is combined with `--add-modules m1,m2,...` with named modules.

So this means `--limit-modules` has no meaning when `ALL-MODULE-PATH` is specified. It will be ignored.

Let `--module-path` `MP` be the observable module set `{a, b, c}`. Let `b` have a dependency on `jdk.net`. `a` and `c` don't have any dependency. Then all of the following are equivalent:

- `jlink --add-modules ALL-MODULE-PATH --module-path MP --limit-modules jdk.net`
- `jlink --add-modules ALL-MODULE-PATH --module-path MP --limit-modules b` (2)
- `jlink --add-modules ALL-MODULE-PATH --module-path MP --limit-modules java.base`
- `jlink --add-modules ALL-MODULE-PATH --module-path MP --limit-modules a`
- `jlink --add-modules ALL-MODULE-PATH --module-path MP --limit-modules c`
- `jlink --add-modules ALL-MODULE-PATH --module-path MP`

Note that one pre-existing test case verifies case (2) above. It asserts that the modules in the image only include the transitive closure of `b` (not including `a`, `c`). This would then change to also include `a` and `c`.

The proposal would essentially mean that `--limit-modules` would be ignored. A corollary of it would be that if the module path includes the `jmods` folder, it would always include all JDK modules plus any that are observable outside that set without any means to cut the list of included modules.

Fine by me, but in that case it should perhaps trigger a warning that `--limit-modules` is being ignored when `ALL-MODULE-PATH` is specified in `--add-modules`.

Is that what you are proposing?

mlchung commented:
The combination of ALL-MODULE-PATH and `--limit-modules` indeed needs time to discuss.  As the tests work with the current fix, we should separate this issue and revisit this in the future.

I have reviewed the version before the commit to check proactively any dependence not observable [bfac446](https://github.com/openjdk/jdk/pull/22494/commits/bfac44659e7219afe472d0e61735df119065796d).   The resolution will fail anyway and this eager verification is not needed IMO.  What do you think?

mlchung commented:
See below as my response to your question but I take back this proposal.  I'm ok to proceed as is and revisit this issue in the future.

> So this means `--limit-modules` has no meaning when `ALL-MODULE-PATH` is specified. It will be ignored.

I can't tell how it will be ignored.  Maybe missing something.

The transitive dependences of the modules specified to `--limit-modules` plus the modules specified to `--add-modules` are observable.   See the implementation of `limitFinder` method which do not compute the dependences of the modules specified to `--add-modules`.

>     * `jlink --add-modules ALL-MODULE-PATH --module-path MP --limit-modules java.base`

This jlink command should fail as `jdk.net` required by `b` is not observable.

> Note that one pre-existing test case verifies case (2) above. It asserts that the modules in the image only include the transitive closure of `b` (not including `a`, `c`). This would then change to also include `a` and `c`.

Yes, I'm aware of that and needs discussion.

> A corollary of it would be that if the module path includes the `jmods` folder, it would always include all JDK modules plus any that are observable outside that set without any means to cut the list of included modules.

JDK modules are not on `--module-path` by default.  So they will not be included.   If `--module-path $JAVA_HOME/jmods` is specified with `--add-modules ALL-MODULE-PATH`, it explicitly requests to add all packaged modules from `jmods` directory.

jerboaa commented:
> > So this means `--limit-modules` has no meaning when `ALL-MODULE-PATH` is specified. It will be ignored.
> 
> I can't tell how it will be ignored. Maybe missing something.
> 
> The transitive dependences of the modules specified to `--limit-modules` plus the modules specified to `--add-modules` are observable. See the implementation of `limitFinder` method which do not compute the dependences of the modules specified to `--add-modules`.
> 
> > ```
> > * `jlink --add-modules ALL-MODULE-PATH --module-path MP --limit-modules java.base`
> > ```
> This jlink command should fail as jdk.net required by b is not observable.

Maybe I'm missing something, too. It seems we have a different understand of *what* the transitive closure is being taken.

`[...] plus the modules specified to '--add-modules' are observable`. So in the above case `MP` has `{a,b,c}` as observable because ALL-MODULE-PATH has been specified. Thus, since I understood it as the "transitive closure of `{a, b, c}` and `java.base`" => `{a, b, jdk.net, c, java.base}` it would not fail.

It sounds to me that you interpret "transitive closure of named modules plus the modules added with --add-modules" as `{java.base, a, b, c}`. Is that it?

Anyway, lets continue this discussion on a separate issue if possible.

mlchung commented:
```
$ jlink --add-modules ALL-MODULE-PATH --module-path build/jmods --limit-modules jdk.net --output ./build/test-img
```

This creates an image with jdk.net and its transitive dependences which is the existing behavior, right?  We are trying to revisit `--add-modules ALL-MODULE-PATH` with `--limit-modules` in the future. 

> "transitive closure of named modules plus the modules added with --add-modules" as {java.base, a, b, c}

Yes.  That's how it is implemented as  https://openjdk.org/jeps/261#Limiting-the-observable-modules describes:

> (The transitive closure computed for the interpretation of the --limit-modules option is a temporary result, used only to compute the limited set of observable modules. The resolver will be invoked again in order to compute the actual module graph.)

So jlink sets up the finder that limits the observability and compute the module graph from the roots using this finder.

AlanBateman commented:
I think it helps to think of `--add-modules ALL-MODULE-PATH` as `--add-modules java.base,java.compiler,java.datatransfer, .. jdk.xml.dom,jdk.zipfs`.

This leads `jlink --add-modules ALL-MODULE-PATH --limit-modules jdk.net` to create an image with "all modules". This may be surprising on first glance but it's the recursive enumeration so `jdk.net` (which is just `jdk.net` and `java.base`), plus the modules specified to `--add-modules`.

Note that we have issues at run-time with this combination, and at run-time there are other possible tokens that can be used too. Not for here, this issue has always existed.

jerboaa commented:
> Only if the `$JAVA_HOME/jmods` folder is also added to the module path. Existing behaviour is like this:

Never mind. That's what JDK-8345573 changed.

I've reverted the eager configuration checking.

mlchung commented:
Now I see better from JDK 23 implementation.  `jlink --add-modules ALL-MODULE-PATH  --module-path mods --limit-modules XXX" always fails.   The reason is that it always fails to create the module finder with a non-empty limit modules for `ALL-MODULE-PATH`.

```
      ModuleFinder finder = newModuleFinder(options.modulePath, options.limitMods, Set.of());
```      

https://github.com/openjdk/jdk23u/blob/master/src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java#L374

jerboaa commented:
Yes. See https://bugs.openjdk.org/browse/JDK-8345573?focusedId=14729247&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14729247

mlchung commented:
This solution is simpler -- it can just detect if `ALL-MODULE-PATH` is used with `--limit-modules`, jlink just fails.

AlanBateman commented:
> This solution is simpler -- it can just detect if `ALL-MODULE-PATH` is used with `--limit-modules`, jlink just fails.

That seems like a good suggestion to remove the complexity and need to think what it means. It would avoid complexity in the other phases too.

jerboaa commented:
The latest update disallows `--limit-modules` when `ALL-MODULE-PATH` is specified. PTAL. Thanks!

mlchung commented on test/jdk/tools/jlink/basic/AllModulePath.java:
```diff
@@ -84,60 +101,135 @@ public void setup() throws Throwable {
+        Path image = HELPER.createNewImageDir("image2");
+        List<String> opts = List.of("--module-path", MODS.toString(),
+                                    "--output", image.toString(),
+                                    "--add-modules", "m1,test",
```
```suggestion
                                    "--add-modules", "m1",
```

This can show clearly that `ALL-MODULE-PATH` includes all modules on the module path.

mlchung reviewed (commented):
Looks good.  Please update the copyright end year.

jerboaa commented on test/jdk/tools/jlink/basic/AllModulePath.java:
```diff
@@ -84,60 +101,135 @@ public void setup() throws Throwable {
     @Test
-    public void testLimitModules() throws Throwable {
-        if (Files.notExists(JMODS)) {
+    public void testSubsetModules() throws Throwable {
```
Sure.

jerboaa commented:
> Looks good. Please update the copyright end year.

Thanks for the review! Done in the latest update.

jerboaa commented:
> Looks good. This PR uncovers unanticipated issues and thanks for looking at them thoroughly.

Thanks for the thorough reviews as well!

jerboaa commented:
@AlanBateman OK for you too? Thanks!

AlanBateman commented:
> @AlanBateman OK for you too? Thanks!

So --limit-modules && --add-modules ALL-MODULE-PATH is an error, as is --add-modules ALL-MODULE-PATH without specifying --module-path. Happy with that. I don't have time to go through the test but Mandy has already reviewed so I think we are good.

jerboaa commented:
> > @AlanBateman OK for you too? Thanks!
> 
> So --limit-modules && --add-modules ALL-MODULE-PATH is an error, as is --add-modules ALL-MODULE-PATH without specifying --module-path. Happy with that. I don't have time to go through the test but Mandy has already reviewed so I think we are good. I'm just wondering if this needs a CSR and also whether the man page needs update.

Thanks. This has got a CSR.

jerboaa commented:
/integrate

jerboaa commented:
> whether the man page needs update.

If you think that's needed then I'm fine to follow up with it. I don't think that's warranted. There will be a release note.

mlchung commented:
The man page does not mention about `ALL-MODULE-PATH`.    As we believe it's rarely used, I'm okay to leave it as is.

```
`--add-modules` *mod*\[`,`*mod*...\]
:   Adds the named modules, *mod*, to the default set of root modules. The
    default set of root modules is empty.
```
